service/shadowsocks.go

```diff
@@ -14,13 +14,128 @@
 
 package service
 
-import "time"
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+)
+
+const (
+	// 59 seconds is most common timeout for servers that do not respond to invalid requests
+	tcpReadTimeout time.Duration = 59 * time.Second
+
+	// A UDP NAT timeout of at least 5 minutes is recommended in RFC 4787 Section 4.3.
+	defaultNatTimeout time.Duration = 5 * time.Minute
+)
 
 // ShadowsocksConnMetrics is used to report Shadowsocks related metrics on connections.
 type ShadowsocksConnMetrics interface {
 	AddCipherSearch(accessKeyFound bool, timeToCipher time.Duration)
 }
 
+type ServiceMetrics interface {
+	UDPMetrics
+	AddOpenTCPConnection(conn net.Conn) TCPConnMetrics
+	AddCipherSearch(proto string, accessKeyFound bool, timeToCipher time.Duration)
+}
+
+type Service interface {
+	HandleStream(ctx context.Context, conn transport.StreamConn)
+	HandlePacket(conn net.PacketConn)
+}
+
+// Option is a Shadowsocks service constructor option.
+type Option func(s *ssService)
+
+type ssService struct {
+	metrics     ServiceMetrics
+	ciphers     CipherList
+	natTimeout  time.Duration
+	replayCache *ReplayCache
```

Comment:
Can the replayCache be nil?

Reply:
Yes, it can. The implementation of `ReplayCache` includes a safeguard against `nil` pointers: https://github.com/Jigsaw-Code/outline-ss-server/blame/4463b3886f09f251c6b15662bcb56b4e1ef26f25/service/replay.go#L83. That's probably why we don't explicitly safeguard against it in `NewShadowsocksStreamAuthenticator()`.

I'm not sure if that pattern is idiomatic in Go?